docs/porting-to-ROS2.md

```diff
@@ -150,7 +150,32 @@ There is a [migration guide](https://index.ros.org/doc/ros2/Tutorials/Launch-fil
 
 
 ### Replacing `tf2_ros::Buffer`
-A `tf2_ros::Buffer` member that is filled by a `tf2_ros::TransformListener` can become a `tf2::BufferCore` instead. For an example, see [this PR](https://github.com/tier4/Pilot.Auto/pull/11)
+A `tf2_ros::Buffer` member that is filled by a `tf2_ros::TransformListener` can become a `tf2::BufferCore` in most cases. This reduces porting effort, since the a `tf2::BufferCore` can be constructed like a ROS1 `tf2_ros::Buffer`. For an example, see [this PR](https://github.com/tier4/Pilot.Auto/pull/11).
+
+However, in some cases the extra functionality of `tf2_ros::Buffer` is needed. For instance, waiting for a transform to arrive.
+
+
+#### Waiting for a transform to arrive
+You might expect to be able to use `tf2_ros::Buffer::lookupTransform()` with a timeout out of the box, but this will throw an error:
+
+    Do not call canTransform or lookupTransform with a timeout unless you are using another thread for populating data. 
+    Without a dedicated thread it will always timeout. 
+    If you have a seperate thread servicing tf messages, call setUsingDedicatedThread(true) on your Buffer instance.
+
+You could do therefore try setting up a dedicated thread, but you could also use the `waitForTransform()` function like this:
+
+    // In the node definition
+    tf2_ros::Buffer tf_buffer_;
+    tf2_ros::TransformListener tf_listener_;
+    ...
+    // In the constructor
+    auto cti = std::make_shared<tf2_ros::CreateTimerROS>(this->get_node_base_interface(), this->get_node_timers_interface());
+    tf_buffer_.setCreateTimerInterface(cti);
+    ...
+    // In the function processing data
+    tf_buffer_.waitForTransform(a, b, msg_time, std::chrono::milliseconds(1000), [this](const std::shared_future<geometry_msgs::msg::TransformStamped>& tf){...}
```

Comment:
I think this last line might actually be incomplete (missing closing brace and semi colon), when I tried to implement this line in my code I had compilation errors. I looked at the implementation used in the ROS2 `tf2_ros::Buffer` test code and used this example as well: https://github.com/ros2/geometry2/blob/foxy/tf2_ros/test/test_buffer.cpp#L140-L158

Reply:
Maybe I don't understand it, but that example seems very wrong to me. The code after the callback continues executing immediately, I checked.

You're correct about the incomplete line, I fixed that. I also added a sentence on the future that is returned by the function.

Reply:
Hmm maybe I don't understand it entirely either. It seems like the `future` object is called again twice: 
* https://github.com/ros2/geometry2/blob/foxy/tf2_ros/test/test_buffer.cpp#L176
* https://github.com/ros2/geometry2/blob/foxy/tf2_ros/test/test_buffer.cpp#L179

Reply:
Ah yes, you're right, the future is again waited on. I didn't see that and thought `transform_callback_result` was accessed with no synchronization at all. In that case, the test you linked is fine, but only for the case where the transform arrives in time. If the transform does not arrive in time, the outer future will always block forever, like described in the docs.

So I think you would need to handle that case in your PR. Also, the exception will not propagate outside `waitForTransform()`, that one will never throw, so `try`/`catch` would only make sense inside the callback.

For how to handle that: I think it's easiest to do everything that currently comes after `waitForTransform()` in the callback. Alternatively, you could `wait_for()` on the outside future so that you won't risk waiting forever, and then check the returned status before calling `get()`. I think it's not very easy for a reader of the code to understand the subtleties of that approach, and it basically does a second timeout of the same duration, so leads to repetition.

Reply:
I think I will follow what you are saying. Let me see if i understand correctly:

```
geometry_msgs::msg::TransformStamped return_result;
auto future = buffer.waitForTransform(
  "foo",
  "bar",
  tf2_time, tf2::durationFromSec(1.0),
  [&return_result](const tf2_ros::TransformStampedFuture & future)
    {
      try {
        transform_callback_result = future.get();
        ...
        // process result
        return_result = something;
      } catch (...) {
        // handle error
      }
    }
);
```

This doesn't seem right to me but maybe I haven't understood it properly. I'm still trying to wrap my head around it. I understand what you mean by it seems to would lead to repetition, it seems like the test code tries to verifies both methods you mentioned before.

Reply:
So, you're capturing `return_result` and assigning to it in the callback, which makes me think you want to use that variable outside the callback after the call to `waitForTransform()`. If you _don't_ use it outside, but only inside where it says `// process result`, then that would be correct, and in that case you don't need to (and shouldn't) create an outer variable and capture it.

If your intent is to do something with `return_result` below `waitForTransform()`, the code sample is not correct yet. Since `waitForTransform()` returns immediately, the result will not have been written to `return_result` yet. So you need to wait for the future or build some other way of signalling that the result has been written to `return_result`.

But the details of that are tricky and I am not 100% done understanding them. I have uploaded a repo for easily testing things yourself [here](https://github.com/nnmm/tf2_example). After you clone it into a workspace and build, run the waiter node first with 

    ros2 run tf_example waiter --ros-args -p timeout_ms:=100 -p wait_inside:=true -p wait_outside:=true

and after that similarly for the publisher which has no parameters. What's important is that the transform requested in `Waiter::topic_callback()` arrives one second after the callback gets called.

So what are the details? With the above timeout (which should expire since 100ms << 1s):
– Waiting inside + outside hits the outer get() which doesn't respect the timeout but just waits the full second, and then returns the transform. After that, the callback runs, which doesn't throw an exception. 
– Waiting only outside is the same
– Waiting only inside, that _does_ change the behavior of the inner call to `get()`: It now correctly throws the exception.

So it seems that the future is able to make progress while we're still in the message callback. That's a bit surprising because my impression was that when you're in a message callback, no other thread is working to receive tf messages. At the same time, the `waitForTransform()` callback cannot be called while we're still in the message callback (I also tested this with a manual `std::this_thread::sleep_for(std::chrono::milliseconds(500))`. Perhaps this behavior is also different with a multithreaded executor, similar to what we saw with clients?

Okay, so `wait()` or `get()` in the outside code is useless at least with a singlethreaded executor. Our only hope is `wait_for()`, which the other test cases are using. Let's try replacing the outside waiting code with

    RCLCPP_INFO(get_logger(), "[outside %d] Calling wait_for()", t);
    auto status = future.wait_for(std::chrono::milliseconds(this-timeout_ms_));
    if (status == std::future_status::deferred) {
      RCLCPP_INFO(get_logger(), "[outside %d] Status is deferred", t);
    } else if (status == std::future_status::timeout) {
      RCLCPP_INFO(get_logger(), "[outside %d] Status is timeout", t);
    } else {
      RCLCPP_INFO(get_logger(), "[outside %d] Status is ready", t);
    }

that seems to work: The status is `timeout` when the transform didn't arrive in the time specified in the `wait_for()` – the timeout specified in `waitForTransform()` doesn't seem to matter, which is also confusing. If the transform arrived, it will be `ready`. An open question is whether it can ever be `deferred`.

Another open question: With this setup, the  `waitForTransform()` callback manages gets called while the message callback is still running (directly after the `wait_for()` returns). I'm not sure why.

I'd go with just ignoring the returned value because it is complicated, and working with the future passed into the callback.

Reply:
Thanks for the detailed explanation and the example. This definitely helps my understanding a lot!

> which makes me think you want to use that variable outside the callback after the call to waitForTransform()

I need to transform it from a message into a tf2 type to be processed downstream. So, yes, the return value is quite crucial. I don't think it can be done without the return value. The test case i linked above captures the return result and then reads the contents further along the execution so it is exactly what I want to achieve. 

> Perhaps this behavior is also different with a multithreaded executor, similar to what we saw with clients? 

I've tried out the example you created. I was thinking about your remark about threading and I think we overlooked something during the construction of the buffer objects: https://github.com/ros2/geometry2/blob/foxy/tf2_ros/test/test_buffer.cpp#L143. However when I tried setting this and checking the output, it seemed to not make any difference. I thought this would put the buffer object into a seperate thread and therefore allow the callback to be executed while still in the message calback?

> that seems to work: The status is timeout when the transform didn't arrive in the time specified in the wait_for() – the timeout specified in waitForTransform() doesn't seem to matter, which is also confusing. If the transform arrived, it will be ready.

In that case, let me see if I can correct my example above:

```
int t = msg->header.stamp.sec;

// No return value
auto future = buffer.waitForTransform(
  "foo",
  "bar",
  tf2_time, std::chrono::milliseconds(this->timeout_ms_),
  [this](const tf2_ros::TransformStampedFuture & future)
    {
      try {
        auto do_not_care = future.get();
      } catch (...) {
        // handle error case (maybe this try catch is no longer necessary if we can handle the exceptions )
      }
    }
  );

// Check the status of future to see if the return value has been written to
RCLCPP_INFO(get_logger(), "[outside %d] Calling wait_for()", t);
auto status = future.wait_for(std::chrono::milliseconds(this-timeout_ms_));
if (status == std::future_status::ready) {
  RCLCPP_INFO(get_logger(), "[outside %d] Status is ready", t);
} else {
  RCLCPP_INFO(get_logger(), "[outside %d] Failed to retrieve transform", t);
  return;
}

// Future can be read now and contains the transform data
auto desired_transform = future.get();

// do things with desired_transform
```

I'm still a little confused as to what should go into the future callback. A try catch seems necessary... but if we are checking the status of `wait_for()` I don't know if it's all that necessary. I'll keep looking into this. 

Another note, I'm not sure if it is worth writing some sort of wrapper around these function calls? I don't know how many other places need this but it would make the code base rather bloated to have to add this many lines (and needless complexity) to replace a single `waitForTransform` call. It can be part of the code clean up, post-porting.

Reply:
Responding to the issues you brought up:
> I don't think it can be done without the return value.

In principle, it should always be possible by refactoring the code that expects a return value to pass in a callback, but it might make the code very ugly in practice in some cases. I did something very similar with a service call [here](https://github.com/tier4/Pilot.Auto/pull/11/files#diff-f1e72b5ed7c18fcba14d8007e32d3d2c4c3e736e03bafd59605063df3167ca62R179-R208) – previously that function had a return value, `output_pose_msg_ptr`, and the caller of `callAlignPose` would give that return value to `initial_pose_pub_->publish()`. Now this happens directly in the callback.

> I thought this would put the buffer object into a separate thread and therefore allow the callback to be executed while still in the message callback?

My impression was that this was just for suppressing the warning and you have to set up the separate thread yourself, but I'll need to get back to you with a source for that.

> I don't quite understand how the get() is useless outside of the callback.

Yes, right, I should have said `wait()`, or `get()` on a not-yet-ready future which will `wait()`. That `wait()`, whether direct or indirect, is useless because it doesn't respect the timeout.

> I'm still a little confused as to what should go into the future callback

If you choose to do the "synchronous" approach, i.e. work with the returned future instead of the callback, what Taka did looks right to me: https://github.com/tier4/Pilot.Auto/blob/42f47dcf36eec421c5ab8e8b63fd320e07e02d59/control/velocity_controller/src/velocity_controller_core.cpp#L184-L192 You then shouldn't need anything in the future callback. The try/catch is replaced by checking the status returned by `wait_for`.

Reply:
Great! Thanks for the explanation I think that has really cleared it up for me now. Sorry about all the questions! Some of the these concepts are new to me.

Reply:
Thanks :) No worries, those questions caused me to take a closer look and learn a lot as well.